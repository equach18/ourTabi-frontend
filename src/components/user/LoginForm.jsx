import React, { useState, useEffect } from "react";
import { useNavigate, useLocation } from "react-router-dom";

import Alert from "../common/Alert";

/** Login form.
 *
 * On submission:
 * - Calls login function prop
 * - Redirects to /dashboard route if successful
 */

function LoginForm({ login }) {
  const navigate = useNavigate();
  const location = useLocation();

  const [formData, setFormData] = useState({
    username: "",
    password: "",
  });
  const [formErrors, setFormErrors] = useState([]);

  useEffect(() => {
    if (location.state?.alert) {
      setFormErrors([location.state.alert]);
    }
  }, [location.state]);

  /** Handle form submit: */
  async function handleSubmit(evt) {
    evt.preventDefault();
    setFormErrors([]);

    let result = await login(formData);
    if (result.success) {
      navigate("/dashboard");
    } else {
      setFormErrors(
        Array.isArray(result.errors) ? result.errors : ["Login failed."]
      );
    }
  }

  /** Update form data field */
  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((l) => ({ ...l, [name]: value }));
  }

  return (
    <div className="flex flex-col items-center justify-center min-h-[calc(100vh-4rem)] bg-gradient-to-r from-stone-300 via-white to-stone-300 px-4">
      <h1 className="text-3xl font-bold text-orange-500 font-heading mb-4">
        Welcome Back
      </h1>

      <form
        onSubmit={handleSubmit}
        className="bg-white p-6 rounded-xl shadow-lg w-full max-w-sm"
      >
        {formErrors.length > 0 && (
          <div className="mb-4">
            <Alert type="error" messages={formErrors} />
          </div>
        )}

        <div className="mb-3">
          <label className="block text-zinc-700 text-sm font-medium mb-1" htmlFor="username">
            Username
          </label>
          <input
            type="text"
            name="username"
            id="username"
            value={formData.username}
            onChange={handleChange}
            className="w-full py-1 px-2 text-sm border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            required
          />
        </div>

        <div className="mb-4">
          <label className="block text-zinc-700 text-sm font-medium mb-1" htmlFor="password">
            Password
          </label>
          <input
            type="password"
            name="password"
            id="password"
            value={formData.password}
            onChange={handleChange}
            className="w-full py-1 px-2 text-sm border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            required
          />
        </div>

        <button
          type="submit"
          className="bg-emerald-600 text-white px-4 py-2 rounded-lg w-full hover:bg-emerald-700 transition"
        >
          Login
        </button>
      </form>
    </div>
  );
}

export default LoginForm;
