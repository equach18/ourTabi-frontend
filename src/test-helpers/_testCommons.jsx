import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserContext from "../context/UserContext";
import TripContext from "../context/TripContext";

export const testUser1 = {
  id: 1,
  firstName: "Test1",
  lastName: "User1",
  username: "testUser1",
  profilePic: "",
};
export const testUser2 = {
  id: 2,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser2",
  profilePic: "testUser2.png",
};
export const testUser3 = {
  id: 3,
  firstName: "Test2",
  lastName: "User2",
  username: "testUser3",
  profilePic: "testUser3.png",
};

export const defaultMembers = [
  {
    userId: testUser1.id,
    username: testUser1.username,
    profilePic: testUser1.profilePic,
  },
  {
    userId: testUser2.id,
    username: testUser2.username,
    profilePic: testUser2.profilePic,
  },
];

export const testActivity = {
  id: 1,
  name: "Kayaking",
  tripId: 53,
  category: "Other",
  location: "Chicago",
  description: "So much fun",
  scheduledTime: "2025-07-10T14:30:00Z",
};

export const testComment = {
  id: 1,
  text: "Test comment",
  createdAt: "2025-07-11T14:30:00Z",
  userId: 1,
};

export const testTrip = {
  id: 42,
  title: "Test trip",
  destination: "Chicago, IL",
  startDate: "2025-07-10",
  endDate: "2025-07-12",
  isPrivate: true,
};

export const testFriendsData = {
  friends: [testUser1 ],
  incomingRequests: [testUser2],
  sentRequests: [testUser3],
};

// Default mock context values
const defaultUserContext = { currentUser: testUser1 };
const defaultTripContext = {
  voteOnActivity: vi.fn(),
  votes: {},
  deleteComment: vi.fn(),
  addMember: vi.fn(),
  removeMember: vi.fn(),
  members: defaultMembers,
};

// create mock trip context
export function createMockTripContext(overrides = {}) {
  return {
    voteOnActivity: vi.fn(),
    votes: {},
    deleteComment: vi.fn(),
    members: [],
    addMember: vi.fn(),
    removeMember: vi.fn(),
    ...overrides,
  };
}

/**
 * Renders a component inside MemoryRouter
 */
export function renderWithRouter(ui) {
  return render(<MemoryRouter>{ui}</MemoryRouter>);
}

/**
 * Renders a component inside MemoryRouter + UserContext + TripContext
 */
export function renderWithAllProviders(
  ui,
  {
    userContext = defaultUserContext,
    tripContext = defaultTripContext,
    route = "/",
  } = {}
) {
  return render(
    <MemoryRouter initialEntries={[route]}>
      <UserContext.Provider value={userContext}>
        <TripContext.Provider value={tripContext}>{ui}</TripContext.Provider>
      </UserContext.Provider>
    </MemoryRouter>
  );
}

/**
 * Renders a component with just UserContext
 */
export function renderWithUserProvider(ui, { userContext } = {}) {
  return render(
    <UserContext.Provider value={userContext}>{ui}</UserContext.Provider>
  );
}
