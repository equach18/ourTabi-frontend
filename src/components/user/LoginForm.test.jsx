import { fireEvent, screen, waitFor } from "@testing-library/react";
import { Routes, Route } from "react-router-dom";
import { renderWithAllProviders } from "../../test-helpers/_testCommons";
import LoginForm from "./LoginForm";

const mockNavigate = vi.fn();

vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockNavigate,
    useLocation: () => ({ state: null }),
  };
});

describe("LoginForm", () => {
  const mockLogin = vi.fn();

  function renderForm() {
    return renderWithAllProviders(
      <Routes>
        <Route path="/login" element={<LoginForm login={mockLogin} />} />
      </Routes>,
      { route: "/login" }
    );
  }

  beforeEach(() => {
    mockLogin.mockReset();
    mockNavigate.mockReset();
  });

  it("matches snapshot", () => {
    const { asFragment } = renderForm();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the login form", () => {
    renderForm();

    expect(screen.getByLabelText(/username/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/password/i)).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /login/i })).toBeInTheDocument();
  });

  it("calls login and navigates to dashboard on success", async () => {
    mockLogin.mockResolvedValue({ success: true });

    renderForm();

    fireEvent.change(screen.getByLabelText(/username/i), {
      target: { value: "testuser" },
    });
    fireEvent.change(screen.getByLabelText(/password/i), {
      target: { value: "password123" },
    });
    fireEvent.click(screen.getByRole("button", { name: /login/i }));

    await waitFor(() => {
      expect(mockLogin).toHaveBeenCalledWith({
        username: "testuser",
        password: "password123",
      });
      expect(mockNavigate).toHaveBeenCalledWith("/dashboard");
    });
  });

  it("shows error msgs when login fails", async () => {
    mockLogin.mockResolvedValue({
      success: false,
      errors: ["Invalid credentials"],
    });

    renderForm();

    fireEvent.change(screen.getByLabelText(/username/i), {
      target: { value: "wrong" },
    });
    fireEvent.change(screen.getByLabelText(/password/i), {
      target: { value: "badpw" },
    });
    fireEvent.click(screen.getByRole("button", { name: /login/i }));

    expect(await screen.findByText(/invalid credentials/i)).toBeInTheDocument();
  });
});
