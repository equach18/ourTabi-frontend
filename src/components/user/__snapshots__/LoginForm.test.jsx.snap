// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col items-center justify-center min-h-[calc(100vh-4rem)] bg-gradient-to-r from-stone-300 via-white to-stone-300 px-4"
  >
    <h1
      class="text-3xl font-bold text-orange-500 font-heading mb-4"
    >
      Welcome Back
    </h1>
    <form
      class="bg-white p-6 rounded-xl shadow-lg w-full max-w-sm"
    >
      <div
        class="mb-3"
      >
        <label
          class="block text-zinc-700 text-sm font-medium mb-1"
          for="username"
        >
          Username
        </label>
        <input
          class="w-full py-1 px-2 text-sm border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
          id="username"
          name="username"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-4"
      >
        <label
          class="block text-zinc-700 text-sm font-medium mb-1"
          for="password"
        >
          Password
        </label>
        <input
          class="w-full py-1 px-2 text-sm border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
          id="password"
          name="password"
          required=""
          type="password"
          value=""
        />
      </div>
      <button
        class="bg-emerald-600 text-white px-4 py-2 rounded-lg w-full hover:bg-emerald-700 transition"
        type="submit"
      >
        Login
      </button>
    </form>
  </div>
</DocumentFragment>
`;
